### ISSUE TITLE
UnnecessaryNotNullCheck: false negative with smart casted arguments

### ISSUE THREAD

**Issue by t-kameyama** [detekt/detekt#5379]
## Steps to Reproduce
```kotlin
 @Test
 fun shouldDetectAfterNullCheck() {
     val code = """
         fun foo(x: Int?) {
             if (x != null) {
                 requireNotNull(x)
             }
         }
     """.trimIndent()
     val findings = subject.lintWithContext(env, code)
     assertThat(findings).hasSize(1)
 }

 @Test
 fun shouldDetectAfterTypeCheck() {
     val code = """
         fun bar(x: Any?) {
             if (x is String) {
                 requireNotNull(x)
             }
         }
     """.trimIndent()
     val findings = subject.lintWithContext(env, code)
     assertThat(findings).hasSize(1)
 }
```
